### PR TITLE
fix cp symlink to './' problem

### LINKED ISSUE #11394: FATA[0000] invalid symlink when copying a symlink with relative parent paths
```Dockerfile
FROM debian:jessie

RUN mkdir -p /tmp/test/subdir1 /tmp/test/subdir2 \
    && touch /tmp/test/subdir1/a \
    && cd /tmp/test/subdir2 && ln -s ../subdir1/a
```

```sh
$docker build -t symlink .
[...]

$ docker run --rm -ti symlink ls -l /tmp/test/subdir2
total 0
lrwxrwxrwx 1 root root 12 Mar 15 13:05 a -> ../subdir1/a

$ docker create --name test symlink
287d64308a8264cf46bdde93cd53dc85abbc107c363736df89c0f6498f83c87c

$ docker cp test:/tmp/test .
FATA[0000] invalid symlink "test/subdir2/a" -> "../subdir1/a"
```

Tested on a Linux host. There was a similar issue (#7296) but it was suggested there, that I should rather create a new issue as the other issue was windows related and thus a different problem than this here.

UPDATE: Fixed the `Dockerfile`.

### PR DESCRIPTION
**Pull request by x1022as**
fix #11394

Signed-off-by: Deng Guangxing dengguangxing@huawei.com

### CODE REVIEW

**Comment by x1022as** [comment]
#11394 failed because the destination is `.` or `./`
in such case, it failes here:
```bash
if strings.HasPrefix(targetPath, extractDir) {
         return breakoutError(fmt.Errorf("invalid symlink %q -> %q", path, hdr.Linkname))
}
```
where extractDir is `'.'` and targetPath does not have such prefix .

**Comment by coolljt0725** [comment]
I meet the same problem when I use docker container to build something and copy the result out to host, seems this fix can solve this. Maybe a test is need

**Comment by unclejack** [comment]
This needs tests and it also need to be verified to ensure it doesn't introduce any security vulnerabilities.

**Comment by x1022as** [comment]
Test failed! Does anyone know the reason?

**Comment by jfrazelle** [comment]
It is a registry issue they are looking into it

**Comment by LK4D4** [comment]
ping @unclejack @jfrazelle 
What about security, should we merge this to 1.6?

**Comment by jfrazelle** [comment]
ya I am a bit nervous about this and seeing as its not a regession i say
dont rush

**Comment by unclejack** [comment]
+1 Let's not rush with this one. We should review this after 1.6.0 is out.

**Comment by LK4D4** [comment]
ping @tiborvass @unclejack 
Any concerns about security?

**Comment by unclejack** [comment]
@LK4D4 I'd wait until 1.6.0 is out so we can properly review this.

**Comment by tiborvass** [comment]
@x1022as thank you very much for your contribution. Would you mind rebasing and using the new `DockerSuite` for the test? If you cannot, we can carry your commit, just let us know! Thanks!

**Comment by x1022as** [comment]
@tiborvass I will rebase and change the test.

**Comment by jfrazelle** [comment]
LGTM ping @tiborvass @unclejack

**Comment by jfrazelle** [comment]
ping @tiborvass @unclejack we should get this in early if we want it just to make sure it doent break anything also ping @tonistiigi if you want to take a look :)

File: integration-cli/docker_cli_cp_test.go [hunk 1]
```diff
@@ -204,6 +204,31 @@ func (s *DockerSuite) TestCpAbsolutePath(c *check.C) {
 
 }
 
+func (s *DockerSuite) TestCpSymlinkToCurrentDir(c *check.C) {
```
Comment (tiborvass) [inline integration-cli/docker_cli_cp_test.go]: @x1022as sorry but the test should not pass on master

**Comment by tonistiigi** [comment]
I pointed out once before that whole this line does not make sense to me. `targetpath` is always absolute so in chrooted environment this check is always false (extractdir is root). For non-chrooted environment this adds limitations that ban extracting absolute and parent symlinks although protection for escapes is in other places.

Absolute symlinks should fail the same way so I guess it would be nice to include it in the test as well.

@tiborvass @unclejack @jfrazelle my e-mail from 26.11.2014 contains a bit more details.

**Comment by tiborvass** [comment]
Thanks for the pointer @tonistiigi I'll investigate

File: integration-cli/docker_cli_cp_test.go (hunk 1 above)
Comment (tonistiigi) [inline integration-cli/docker_cli_cp_test.go]: Probably a race between shell executing commands and `docker cp` already running.

**Comment by thaJeztah** [comment]
ping @tiborvass @unclejack

**Comment by LK4D4** [comment]
@x1022as @tiborvass What's up here? It's in milestone btw.

**Comment by icecrime** [comment]
I'm tentatively dropping from milestone, please yell loud (and soon!) if you think it's a mistake.

**Comment by thaJeztah** [comment]
@icecrime would it be a plan to create a 1.8 milestone for issues that *just* didn't make it?

**Comment by icecrime** [comment]
SGTM! Feel free to do so :-)

**Comment by icecrime** [comment]
*Collective review with @jfrazelle @crosbymichael @calavera*

@tonistiigi It's your call on this one: do you think we should proceed or close? Thanks!

**Comment by tonistiigi** [comment]
Security wise this seems quite harmless but NOT LGTM because I don't think this solves the right problem.

As far as I can see the issue isn't that `.` destination path is somewhat special and can't be used for prefix checking. The actual issue is that when `archive` package extracts into a directory it should never forbid inner symlinks just because they point outside the target dir. Instead it must make sure that if such symlinks exists, they can't be used by some other files to break out of the target dir. Related issue is that currently this check itself is buggy(for chroot and absolute links) and that is why this issue does not appear in other places.

For example, this testcase that fails even with this patch:

```
mkdir -p /tmp/test
ln -s ../../ /tmp/test/to-grandparent


docker cp test:/tmp/test foo
```

If we close this we should open a new issue with proper milestone. There is definitely a bug in here.

**Comment by icecrime** [comment]
Thanks @tonistiigi! And sorry @x1022as, closing this (but the issue remains, so feel free to give it another look in light of the above comments).

**Comment by rubenanapu** [comment]
This issue seems to be back on `Docker version 18.09.7, build 2d0083d`, `Ubuntu 18.04.1 LTS`.

**Comment by thaJeztah** [comment]
@rubenanapu can you open a new issue with steps to reproduce?

**Pull request closed without merging**